src/main/scala/io/iohk/ethereum/utils/RLP.scala

```diff
@@ -0,0 +1,274 @@
+    if ((singleShort & 0xFF) == singleShort) encodeByte(singleShort.toByte)
+    else Array[Byte]((singleShort >> 8 & 0xFF).toByte, (singleShort >> 0 & 0xFF).toByte)
+  }
+
+  def encodeInt(singleInt: Int): Array[Byte] = {
+    singleInt match {
+      case value if value == (value & 0xFF) => encodeByte(singleInt.toByte)
+      case value if value == (value & 0xFFFF) => encodeShort(singleInt.toShort)
+      case value if value == (value & 0xFFFFFF) => Array((singleInt >>> 16).toByte, (singleInt >>> 8).toByte, singleInt.toByte)
+      case value => Array((singleInt >>> 24).toByte, (singleInt >>> 16).toByte, (singleInt >>> 8).toByte, singleInt.toByte)
+    }
+  }
+
+  def decodeInt(bytes: Array[Byte]): Int = {
+    (bytes.length: @switch) match {
+      case 0 => 0: Short
+      case 1 => bytes(0) & 0xFF
+      case 2 => ((bytes(0) & 0xFF) << 8) + (bytes(1) & 0xFF)
+      case 3 => ((bytes(0) & 0xFF) << 16) + ((bytes(1) & 0xFF) << 8) + (bytes(2) & 0xFF)
+      case 4 => ((bytes(0) & 0xFF) << 24) + ((bytes(1) & 0xFF) << 16) + ((bytes(2) & 0xFF) << 8) + (bytes(3) & 0xFF)
+      case _ => throw new Exception("Bytes don't represent an int")
+    }
+  }
+
+  /**
+    * Converts a int value into a byte array.
+    *
+    * @param value - int value to convert
+    * @return value with leading byte that are zeroes striped
+    */
+  private def intToBytesNoLeadZeroes(value: Int): Array[Byte] = ByteBuffer.allocate(4).putInt(value).array().dropWhile(_ == (0: Byte))
+
+  /**
+    * Integer limitation goes up to 2&#94;31-1 so length can never be bigger than MAX_ITEM_LENGTH
+    **/
+  private def encodeLength(length: Int, offset: Int): Try[Array[Byte]] = {
+    length match {
+      case l if l < SizeThreshold => Success(Array((length + offset).toByte))
+      case l if l < MaxItemLength && length > 0xFF =>
+        val binaryLength: Array[Byte] = intToBytesNoLeadZeroes(length)
+        Success((binaryLength.length + offset + SizeThreshold - 1).toByte +: binaryLength)
+      case l if l < MaxItemLength && length < 0xFF => Success(Array((1 + offset + SizeThreshold - 1).toByte, length.toByte))
+      case _ => Failure(new RuntimeException("Input too long"))
+    }
+  }
+
+
+  private def decodeWithPos(data: Array[Byte], pos: Int): Try[(RLPEncodeable, Int)] = Try {
+    if (data.length < 1) (
+      new RLPValue {
+        override def toBytes: Array[Byte] = Array.emptyByteArray
+      }, pos)
+    else {
+      val prefix: Int = data(pos) & 0xFF
+      prefix match {
+        case p if p == OffsetShortItem => (new RLPValue {
+          override def toBytes: Array[Byte] = Array.emptyByteArray
+        }, pos + 1)
+        case p if p < OffsetShortItem => (new RLPValue {
+          override def toBytes: Array[Byte] = Array(data(pos))
+        }, pos + 1)
+        case p if p <= OffsetLongItem =>
+          val length = p - OffsetShortItem
+          val res: Array[Byte] = new Array[Byte](length)
+          Array.copy(data, pos + 1, res, 0, length)
```

Comment:
maybe we can use `ByteString` here to to slice consecutive parts from `data`?
maybe than we could commit position and only pass reminder to recurrent call, sth like:
`val (res, reminder) = data.splitAt(pos+1)` ?